crates/bevy_ecs/src/query/point_query.rs

```diff
@@ -0,0 +1,186 @@
+    ///
+    /// In case of a nonexisting entity or mismatched component, a [`QueryEntityError`] is returned instead.
+    ///
+    /// This is always guaranteed to run in `O(1)` time.
+    #[inline]
+    pub fn get(&mut self, entity: Entity) -> Result<D::Item<'_>, QueryEntityError> {
+        // SAFETY: system runs without conflicts with other systems.
+        // same-system queries have runtime borrow checks when they conflict
+        unsafe { std::mem::transmute(self.get_unsafe(entity)) }
+    }
+
+    /// Returns the query item for the given [`Entity`] without any check.
+    ///
+    /// WARNING: This method should only be recommended in extreme cases where performance is the highest priority.
+    ///
+    /// For a safe alternative see [`Self::get`].
+    ///
+    /// # Safety
+    ///
+    /// `entity` must be valid and not in a pending state.
+    /// `entity` must be within the same `World` which the `Query` was generated.
+    /// `entity` must match the scope of `Query`.
+    #[inline]
+    pub unsafe fn get_unchecked(&mut self, entity: Entity) -> D::Item<'_> {
+        let location = self.entities.get(entity).debug_checked_unwrap();
+        if Self::IS_DENSE {
+            if self.last_storage_id.table_id != location.table_id {
+                // SAFETY: `fetch_state`/`filter_state` are the states that `fetch/filter` were initialized with
+                // `table` is from the world that `fetch/filter` were created for,
+                // `fetch_state`/`filter_state` are the states that `fetch/filter` were initialized with
+                let table = self.tables.get(location.table_id).debug_checked_unwrap();
+                D::set_table(&mut self.fetch, &self.state.fetch_state, table);
+                F::set_table(&mut self.filter, &self.state.filter_state, table);
+                self.last_storage_id.table_id = location.table_id;
+            }
+        } else if self.last_storage_id.archetype_id != location.archetype_id {
+            // SAFETY: `archetype` is from the world that `fetch/filter` were created for,
+            // `fetch_state`/`filter_state` are the states that `fetch/filter` were initialized with
+            // `table` is from the world that `fetch/filter` were created for,
+            // `fetch_state`/`filter_state` are the states that `fetch/filter` were initialized with
+            let archetype = self
+                .archetypes
+                .get(location.archetype_id)
+                .debug_checked_unwrap();
+            let table = self.tables.get(location.table_id).debug_checked_unwrap();
+            D::set_archetype(&mut self.fetch, &self.state.fetch_state, archetype, table);
+            F::set_archetype(&mut self.filter, &self.state.filter_state, archetype, table);
+            self.last_storage_id.archetype_id = location.archetype_id;
+        }
+
+        let item = D::fetch(&mut self.fetch, entity, location.table_row);
+        std::mem::transmute(item)
```

Comment:
```suggestion
        D::shrink(item)
```

We can use shrink here as well, instead of the transmute.